magenta/music/musicxml_parser.py

```diff
@@ -0,0 +1,678 @@
+into tensorflow.magenta.NoteSequence.
+"""
+
+# Imports
+# Python 2 uses integer division for integers. Using this gives the Python 3
+# behavior of producing a float when dividing integers
+from __future__ import division
+
+import xml.etree.ElementTree as ET
+from fractions import Fraction
+from zipfile import ZipFile
+from magenta.music import constants
+
+
+class MusicXMLParserState(object):
+  """Maintains internal state of the MusicXML parser"""
+  def __init__(self):
+    # Default to one division per measure
+    self.divisions = 1
+
+    # Default to a tempo of 120 beats per minute
+    # (This is the default tempo according to the
+    # Standard MIDI Files 1.0 Specification)
+    self.tempo = 120
+
+    # Duration of a single beat in seconds
+    self.seconds_per_beat = 0.5
+
+    # Running total of time for the current event.
+    # Resets to 0 on every part. Affected by <forward> and <backup> elements
+    self.time_position = 0
+
+    # Default to a MIDI velocity of 64 (mf)
+    self.velocity = 64
+
+    # Default MIDI program (1 = grand piano)
+    self.midi_program = 1
+
+    # Current MIDI channel (usually equal to the part number)
+    self.midi_channel = 1
+
+    # Keep track of previous note to get chord timing correct
+    self.previous_note = None
+
+    # Keep track of current transposition level
+    self.transpose = 0
+
+
+class MusicXMLDocument(object):
+  """Internal representation of a MusicXML Document
+
+  Represents the top level object which holds the MusicXML document
+  Responsible for loading the .xml or .mxl file using the getscore method
+  If the file is .mxl, this class uncompresses it
+
+  After the file is loaded, this class then parses the document into memory
+  using the parse method.
+  """
+  def __init__(self, filename):
+    self.filename = filename
+    self.score = self.getscore(filename)
+    self.parts = []
+    self.score_parts = []
+    self.midi_resolution = constants.STANDARD_PPQ
+    self.state = MusicXMLParserState()
+    # Total time in seconds
+    self.total_time = 0
+    self.parse()
+
+  @staticmethod
+  def getscore(filename):
+    """Given a MusicXML file, return the score as an xml.etree.ElementTree
+
+    Given a MusicXML file, return the score as an xml.etree.ElementTree
+    If the file is compress (ends in .mxl), uncompress it first
+
+    Args:
+        filename: The path of a MusicXML file
+    """
+    if filename.endswith('.mxl'):
+      # Compressed MXL file. Uncompress in memory.
+      filename = ZipFile(filename)
+
+      # Ignore files in the META-INF directory.
+      # Only retrieve the first file
+      namelist = filename.namelist()
+      files = [name for name in namelist if not name.startswith('META-INF/')]
+      compressed_file_name = files[0]
+      score = ET.fromstring(filename.read(compressed_file_name))
+    else:
+      # Uncompressed XML file.
+      tree = ET.parse(filename)
+      score = tree.getroot()
+
+    return score
+
+  def parse(self):
```

Comment:
I assume you mean prepend with double underscores. This is done and available in commit 62da424